cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp

```diff
@@ -89,6 +89,88 @@ std::vector<NT::fr> calculate_contract_leaves(BaseRollupInputs baseRollupInputs)
     return contract_leaves;
 }
 
+template <size_t N>
+NT::fr iterate_through_tree_via_sibling_path(NT::fr leaf, NT::uint32 leafIndex, std::array<NT::fr, N> siblingPath)
+{
+    for (size_t i = 0; i < siblingPath.size(); i++) {
+        if (leafIndex & (1 << i)) {
+            leaf = crypto::pedersen_hash::hash_multiple({ leaf, siblingPath[i] });
+        } else {
+            leaf = crypto::pedersen_hash::hash_multiple({ siblingPath[i], leaf });
+        }
+    }
+    return leaf;
+}
+
+template <size_t N>
+void check_merkle_membership(NT::fr leaf, NT::uint32 leafIndex, std::array<NT::fr, N> siblingPath, NT::fr root)
+{
+    auto calculatedRoot = iterate_through_tree_via_sibling_path(leaf, leafIndex, siblingPath);
+    if (calculatedRoot != root) {
+        // throw std::runtime_error("Merkle membership check failed");
+    }
+}
+
+template <size_t N>
+void check_membership_of_subtree_in_snapshot(NT::uint32 depth,
+                                             NT::uint32 nextAvailableLeafIndex,
+                                             std::array<NT::fr, N> siblingPath,
+                                             NT::fr snapshotRoot)
+{
+    stdlib::merkle_tree::MemoryTree empty_subtree = stdlib::merkle_tree::MemoryTree(depth);
+    auto leafToCheck = empty_subtree.root();
+    // next_available_leaf_index is at the leaf level. We need at the subtree level (say height 3). So divide by 8.
+    // (if leaf is at index x, its parent is at index floor(x/2))
+    auto leafIndex = nextAvailableLeafIndex / (NT::uint32(1) << depth);
+
+    check_merkle_membership(leafToCheck, leafIndex, siblingPath, snapshotRoot);
+}
+
+AppendOnlySnapshot insert_subtree_to_private_data_tree(BaseRollupInputs baseRollupInputs,
+                                                       NT::fr new_private_data_subtree_root)
+{
+    // 8 commitments added -> subtree of height 3.
+    // So to calculate the new root, we need to start hashing root of subtree with sibling path[2] onwards.
+    auto siblingPath = baseRollupInputs.new_commitments_subtree_sibling_path;
+    auto leafIndexToInsertAt = baseRollupInputs.start_private_data_tree_snapshot.next_available_leaf_index;
+    auto newTreeSnapshot = baseRollupInputs.start_private_data_tree_snapshot;
+
+    // TODO: Sanity check len of siblingPath > height of subtree
+    // TODO: Ensure height of subtree is 3 (or 8 commitments)
+
+    // if leaf is at index 8, parent is at index floor(x/2)
```

Comment:
The comment here is a little weird for the below computation. Also, use the constants were possible.